deps/happy-eyeballs/happy-eyeballs.c

```diff
@@ -0,0 +1,683 @@
+	 */
+	socklen_t winner_addr_len;
+
+	/**
+	 * error will be non-zero in case of an error during operation.
+	 */
+	int error;
+
+	/**
+	 * error_message may be set when there is a non-zero error code.
+	 */
+	char *error_message;
+
+	/**
+	 * Set along with bind_addr to hint which interface to use.
+	 */
+	socklen_t bind_addr_len;
+
+	/**
+	 * Set along with bind_addr_len to hint which interface to use.
+	 */
+	struct sockaddr_storage bind_addr;
+
+	/**
+	 *  List of in-flight connection attempts.
+	 */
+	DARRAY(struct happy_eyeballs_candidate) candidates;
+
+	/**
+	 *  Protects against multiple simultaneous winners writing socket_fd, winner_addr, and winner_addr_len
+	 */
+	pthread_mutex_t winner_mutex;
+
+	/**
+	 *	Protects against mutating while iterating the candidate list
+	 */
+	pthread_mutex_t candidate_mutex;
+
+	/**
+	 *  Event that signals completion of the race, either via winner or error
+	 */
+	os_event_t *race_completed_event;
+
+	/**
+	 *  Addresses gathered by getaddrinfo
+	 */
+	struct addrinfo *addresses;
+
+	/**
+	 *  Domain name resolution time, in nanoseconds (0 until resolution success)
+	 */
+	uint64_t name_resolution_time_ns;
+
+	/**
+	 *  Connection time, in nanoseconds
+	 */
+	uint64_t connection_time_start;
+
+	uint64_t connection_time_end;
+
+	/**
+	 *  Indicates whether we are in the initial phase of dispatching connections, 
+	 *  or if we are waiting for things to connect or time out.
+	 */
+	volatile bool is_starting;
+};
+
+struct happy_connect_woker_args {
+	int sockfd;
+	struct happy_eyeballs_ctx *context;
+	struct happy_eyeballs_candidate *candidate;
+	struct addrinfo *address;
+};
+
+static int check_comodo(struct happy_eyeballs_ctx *context)
+{
+#ifdef _WIN32
+	//COMODO security software sandbox blocks all DNS by returning "host not found"
+	HOSTENT *h = gethostbyname("localhost");
+	if (!h && GetLastError() == WSAHOST_NOT_FOUND) {
+		context->error = WSAHOST_NOT_FOUND;
+		context->error_message =
+			"happy-eyeballs: Connection test failed. This error is likely caused by Comodo Internet Security running OBS in sandbox mode. Please add OBS to the Comodo automatic sandbox exclusion list, restart OBS and try again (11001).";
+		return STATUS_FAILURE;
+	}
+#else
+	(void)context;
+#endif
+	return STATUS_SUCCESS;
+}
+
+static int build_addr_list(const char *hostname, int port,
+			   struct happy_eyeballs_ctx *context)
+{
+	socklen_t addrlen_hint = 0;
+	if (context->bind_addr_len > 0)
+		addrlen_hint = context->bind_addr_len;
+
+	struct addrinfo hints = {0};
+
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_protocol = IPPROTO_TCP;
+	hints.ai_addrlen = addrlen_hint;
+
+	struct dstr port_str;
+	dstr_init(&port_str);
+	dstr_printf(&port_str, "%d", port);
+	
+	uint64_t start_time = os_gettime_ns();
+	int err = getaddrinfo(hostname, port_str.array, &hints,
+			      &context->addresses);
+	context->name_resolution_time_ns = os_gettime_ns() - start_time;
+	
```

Comment:
```suggestion

	struct dstr port_str;
	dstr_init(&port_str);
	dstr_printf(&port_str, "%d", port);
	
	uint64_t start_time = os_gettime_ns();
	int err = getaddrinfo(hostname, port_str.array, &hints,
			      &context->addresses);
	context->name_resolution_time_ns = os_gettime_ns() - start_time;
	
	dstr_free(&port_str)
```
missing `dstr_free`? Not sure why github shows more differences in the comment (preview)